src/Facility.CodeGen.CSharp/CSharpGenerator.cs

```diff
@@ -356,11 +388,108 @@ private CodeGenFile GenerateDto(ServiceDtoInfo dtoInfo, Context context)
 									}
 								}
 
-								if (validatableFields.Count != 0)
+								if (validateFields.Count != 0)
+								{
+									code.WriteLineSkipOnce();
+									foreach (var fieldInfo in validateFields)
+									{
+										var type = context.GetFieldType(fieldInfo);
+										var propertyName = context.GetFieldPropertyName(fieldInfo);
+										var validation = fieldInfo.Validation!;
+
+										switch (type.Kind)
+										{
+											case ServiceTypeKind.Enum:
+											{
+												code.WriteLine($"if ({propertyName} != null && !{propertyName}.Value.IsDefined())");
+												using (code.Indent())
+													code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{propertyName}\", \"Must be an expected enum value.\");");
+
+												break;
+											}
+
+											case ServiceTypeKind.String:
+											{
+												var validRange = validation.LengthRange;
+												if (validRange != null)
+												{
+													if (validRange.Minimum != null)
+													{
+														code.WriteLine($"if ({propertyName} != null && {propertyName}.Length < {validRange.Minimum})");
+														using (code.Indent())
+															code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", \"Must be at least {validRange.Minimum}\");");
+													}
+
+													if (validRange.Maximum != null)
+													{
+														code.WriteLine($"if ({propertyName} != null && {propertyName}.Length > {validRange.Maximum})");
+														using (code.Indent())
+															code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", \"Must be at most {validRange.Maximum}\");");
+													}
+												}
+
+												var validPattern = validation.RegexPattern;
+												if (validPattern != null)
+												{
+													var regexField = $"s_Valid{propertyName}Pattern";
+													code.WriteLine($"if ({propertyName} != null && !{regexField}.IsMatch({propertyName}))");
+													using (code.Indent())
+														code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{propertyName}\", $\"Must match {{{regexField}}}.\");");
```

Comment:
```suggestion
														code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{propertyName}\", $\"Must match regular expression: {regexField}\");");

```

Brackets feel like an unusual delimiter for a regular expression.